### ISSUE TITLE
Allow more than 16 colors for the new style builtins

### ISSUE THREAD

**Issue by fehnomenal** [elves/elvish#687]
As discussed in #520 the new style builtins only support the 16 named colors from https://en.wikipedia.org/wiki/ANSI_escape_code#3/4_bit.

To get rid of this restriction I propose to internally use the `color.Color` (or a custom) interface. Then it will be the job of the rendering component (web or shell) to display the text properly.

To do so I think there needs a new builtin `color` that can be used by users to create new colors which in turn can be passed to `&fg-color` and `&bg-color` of `styled-segment`. So far so easy but the implementation has a few hairy details.

1. We could go the simple route and just let the user do `color $r $g $b` to get a nice struct implementing `color.Color`. The shell renderer would build the sequences `38;2;<r>;<g>;<b>` or `48;2;<r>;<g>;<b>` and the web renderer something like `<span style="color: #<r><g><b>"/>`. Nice.
But what happens if the terminal does not support 24-bit colors or the user of `color` wants his colored strings to fit into the shell theme (i.e. using only the 16 named colors)? Can the former even be detected by us?

2. We could make `color` a multi function that accepts different arguments based on an option and their count and types. It could support the former `color $r $g $b`, `color $color-name`, `color &force-shell=4bit $color-code` and `color &force-shell=8bit $color-code` all returning `color.Color`s implemented by different structs. The shell variant structs should implement an interface `ShellColorer` that knows how to represent itself as an ANSI sequence. This way the shell renderer can test the type and either calls `ShellColorString()` or builds the sequence as in 1. `ShellColorer` should also force the implementation of `color.Color` so the web renderer has no problems.
Here comes the next trouble: which of the different low-bit color variants should the web display?
Also as pointed out by @iwoloschin [here](https://github.com/elves/elvish/issues/520#issuecomment-392737510) there is no real standard for named colors.

Now that I have written down my thoughts option 2 doesn't even sound that bad with the exception of the color name problematic. Probably the API of `color` can be a bit simpler.

To conveniently use the colors there would be the need to change the style transformers for `styled`. But which `color.Color` is to use here?
Maybe it makes sense to create a color style transformer builder that takes a `color.Color` and returns a style transformer that modifies either the foreground or background color depending on an option.
Or `color` could be extended by that option and directly returns a correct style transformer.

Any comments or suggestions?

**Comment by zzamboni** [comment]
@fehnomenal your ideas sound good to me. By judicious use of default values, backward-compatibility could even be kept. For example, if the second argument to `styled` is a string, assume it's a color name like now, and automatically call `color` on it. If it's a `color` object, just use it as-is.

About the names, I would go with the "standard" [Web Colors](https://en.wikipedia.org/wiki/Web_colors), which incorporates most X11 colors and are understood by most browsers.

**Comment by zzamboni** [comment]
@fehnomenal @xiaq one more question about this: why so much focus on web compatibility for the styles and colors? I can only imagine it's for `elvish -web`, but I wonder if there's a good use case for it. I'm legitimately curious, as I've never used it other than to see how it works.

My gut feeling for a shell would be to focus on compatiblity with as many terminal types/programs as possible, maybe, but not so much with web standards.

Am I missing something? Thanks for your feedback.

**Comment by xiaq** [comment]
@zzamboni Yes and no. I want `elvish -web` to be usable at some moment, and I don't see how this conflicts with having compatibility with terminals.

My idea with colors is that we want a format that

* can represent colors on both terminals and on web

* is reasonably abstract (no low-level escape codes), but still maps to actual presentations in a straightforward way.

About the design:

1. We should **not** mix up color spaces that are not compatible. For instance, the 16-color palette is a user-customizable palette and do not have fixed values, so it should have distinct representations than the other color spaces - the 256 color palette in the terminal, "true color" in the terminal and `#rrggbb` in CSS, all have well-defined values. Of those, we can likely conflate the terminal true-color with CSS RGB colors and use the same representation for both.

2. Since we control the format, name conflicts of terminal colors and CSS colors is not really a problem. For instance, we can simply require CSS color names to be prefixed with `css-` (e.g. `css-red`), or vice versa, require terminal color names to be prefixed with `tty-` (e.g. `tty-red`).

About the internal representation, I think a string representation would already be sufficient. It is easier to use and shouldn't be too hard to parse (we already have a mini-DSL for style transformers...). The easiest way to achieve design goal 1 is to simply have a format for each color space and only conflate the ones that are really exactly the same. Some ideas:

* `default` for reverting to, well, default

* Terminal color names: `red`, `green`, `lightred`, `lightgreen`, ... (this is what we have now)

* 256-color palette: `color0`, `color1`, ... `color255` (I am not quite sure about this one. Using `color` for something as specific as the 256-color palette feels inaccurate.)

* 24-bit color on terminal, or RGB color in CSS: `#rrggbb`

On a side note, it seems to make sense to put `styled` and `styled-segment` in a dedicated builtin `styled:` module, so they will be known as `styled:text` and `styled:segment`. We can keep, for instance, a list of the 16 terminal color names as `$styled:tty-colors`.

It might also make sense to rethink about the format of style transformers and make them a bit more formal. The current set of transformers is a pretty small set of `dash-separated-words`, but `bg-#aabbcc` doesn't really keep the pattern. Maybe something like `bg=#aabbcc` would make more sense and transformers all look like `$attr=$value`. I haven't thought about the boolean attributes yet though. It's another discussion anyway.

**Comment by zzamboni** [comment]
@xiaq thanks for the detailed response. The high-level picture now makes a lot more sense. I'm still not convinced of the need for CSS compatiblity (I still don't see the use case), but I agree that, from a design point of view, it makes sense to make things as generic and abstract and decoupled as possible.

I have a few comments on some of the points you make.

- Yes to the `styled:` module - and having the list of valid TTY colors as a list could be awesome. Then we won't all have to do [this](https://github.com/zzamboni/elvish-themes/blob/use-new-styled/chain.elv#L59) ;)
- 256-color palette names: how about `tty0`, `tty1`, ...?
- About the transformers: how about enabling them through options. Options can receive values, but also default to `$true` when you just specify their name. So for example, you could have:
```
styled:transformers = [
  fg-color = [obj val]{ <change $obj's fg-color to $val> }
  bg-color = [obj val][ <change $obj's bg-color to $val> }
  ...
]
```
This way, `styled:segment text` could simply return the text, but `styled:segment text &fg-color=red &blink` could do the right thing. And this would allow users to create new transformers by adding functions to the map (as long as `styled:segment` can recognize those options dynamically - maybe #584 is needed for this?).

**Comment by xiaq** [comment]
Calling colors in the 256-color palette `tty$number` is an interesting idea. Originally I proposed using `tty` for the 16-color palette. But since the first 16 colors in the 256-color palette are actually just the "standard 16 colors" (see https://en.wikipedia.org/wiki/ANSI_escape_code#8-bit), we can probably just coalesce them into "the" terminal palette. The 24-bit color palette can use the `#rrggbb` syntax.

There really isn't anything special about CSS color compatibility, and I don't see how it conflicts with needs in the terminal. Keeping extensibility in mind doesn't take anything away from you. In fact, if you want support for 24-bit colors *in the terminal*, you also need to decide a way to write those colors.

Re. transformers:

First of all, no user-modifiable map for custom transformers. :) It will be hard to prevent collision from different authors as it is effectively a global namespace. Use a function instead.

That point aside, it is still possible for us to use the option syntax for the builtin transformers like fg color, bg color, etc.

it's a neat idea to use the option syntax for transformers (I think you mean `styled:text` in your examples in the last paragraph). I am not sure about this one. On the one hand it does give you some consistency between `styled:segment` and `styled:text`, but on the other hand, transformers can do more things than just changing some attribute of every segment, like toggling a flag (would that be `&bold=toggle`?). Things that are similar but not *quite* the same can get really confusing.

**Comment by xiaq** [comment]
256-color and 24-bit color are now implemented.
